Xdev Handbook (#7)
ncar-xdev/ncar-xdev.github.io: handbook/more-github.md

# GitHub Glossary

> Knowing how to *use* GitHub on a day-to-day basis is one thing (see [How we use GitHub](github)), but knowing all of the necessary components is another.  Most of the components are *mentioned* in the [How we use GitHub](github)) section, but it's worth while to list the components in more explicit detail so you have a "lay of the land."

Reviewer: ```suggestion
Knowing how to *use* GitHub on a day-to-day basis is one thing (see [How we use GitHub](github)), but knowing all of the necessary components is another.  Most of the components are *mentioned* in the [How we use GitHub](github)) section, but it's worthwhile to list the components in more explicit detail so you have a "lay of the land."
```

## Organization

> To keep us from being to scattered, we have opted to do all of our work (or as much as possible) in our own GitHub organization: [ncar-xdev](https://github.com/ncar-xdev).  If you are a core member of the Xdev Team, you should already be a member (or owner) of the GitHub organization.  If you are a collaborator working with Xdev on a specific project, you should request to be added to the project repositories, as needed.

Reviewer: ```suggestion
To keep us from being too scattered, we have opted to do all of our work (or as much as possible) in our own GitHub organization: [ncar-xdev](https://github.com/ncar-xdev).  If you are a core member of the Xdev Team, you should already be a member (or owner) of the GitHub organization.  If you are a collaborator working with Xdev on a specific project, you should request to be added to the project repositories, as needed.
```

:::{admonition} Tip
:class: tip
You don't have to be a member of a repository's organization to create a fork of that repository *as long as it is public*!  That also means you do not have to be a member of the organization to submit a Pull Request!  You just won't be able to merge the Pull Request yourself.
:::

## Team

As part of our GitHub organization, we are all on a GitHub Team: `@ncar-xdev/all`.  You can use this team name to @mention the entire team on Issues or Pull Request conversations to draw their attention and to indicate that your attention is important.  More specific teams can be created for individual repositories or projects, but this is the "catch-all" team for everything we do.

:::{admonition} Quick Link
:class: note
[Xdev Team](https://github.com/orgs/ncar-xdev/teams/all)
:::

## Discussions Page

> Through this GitHub Team, we also engage in [Team Discussions](https://github.com/orgs/ncar-xdev/teams/all/discussions).  The Team Discussions page should be used every day to share with the rest of the team *what you are planning to do that day.*  This is, in essense, an asynchronous stand-up meeting, and it is a good way for the team to keep in touch with each other and to build an understanding of what we are doing on any given day.

Reviewer: ```suggestion
Through this GitHub Team, we also engage in [Team Discussions](https://github.com/orgs/ncar-xdev/teams/all/discussions).  The Team Discussions page should be used every day to share with the rest of the team *what you are planning to do that day.*  This is, in essence, an asynchronous stand-up meeting, and it is a good way for the team to keep in touch with each other and to build an understanding of what we are doing on any given day.
```

While each project repository can vary depending upon the project, it is safe to assume that each repository has four different kinds of issues:

- **High Priority:** This means that the issue is extremely important, possibly blocking of other development, and that it should be prioritized if at all possible.  These issues are labeled `high priority`.
- **Low Priority:** This means that you shouldn't bother with this right now.  There are other more important things to consider unless you know that addressing this issue is incredibly easy or if there is nothing else to do.  These issues are labeled `low-priority`.
- **Bug Report:** This means there is a bug (incorrect behavior) in the project code.  If the project repository is a *meta repository*, and the actual code is in another repository, then the Issue in the project repository should point to an actual issue in the code repository.  These issues are labeled `bug`.

> - **Enhancement Proposal:** These Issues are for new features that should be added for the project.  They may point to Issues in other repositories.  These Issues are labeled `enhancement`.

Reviewer: ```suggestion
- **Enhancement Proposal:** These Issues are for new features that should be added to the project.  They may point to Issues in other repositories.  These Issues are labeled `enhancement`.
```

## Issues

Issues on repositories represent *tasks* that need to be done (or are done, if they are closed Issues).  When you choose to work on addressing a given Issue, you should *assign yourself* to the Issue to indicate that you will be working on it.  If someone else has already assigned themselves to an Issue you want to work on, don't worry!  You can collaborate.  Refer to the [Xdevdome Rule](github#xdevdome-two-devs-enter-one-team-leaves) for how to deal with this kind of situation.

> :::{admonition} Label issues appropriately!
> :class: warning
> Please label all new issue appropriately when they are created.  This helps the team decide what they need to work on next.  If you need help deciding an Issue's priority, ping the entire [team](#team) on a comment in the Issue to ask for a decision.
> :::

Reviewer: ```suggestion
:::{admonition} Label issues appropriately!
:class: warning
Please label all new issues appropriately when they are created.  This helps the team decide what they need to work on next.  If you need help deciding an Issue's priority, ping the entire [team](#team) on a comment in the Issue to ask for a decision.
:::
```

Remember that once you realize that something needs to be done in a project or on a specific repository, you should create an issue for that task in the appropriate repository so that others might be able to help address it.

## Pull Requests

Pull Requests on repositories represent *tasks* that are being worked on *right now*.  Every Pull Request should reference and be connected to a given [Issue](#issues).  A single Pull Request need not close an Issue, as multiple Pull Requests might need to be completed before an Issue is completely resolved.  Usually, a single Pull Request will not close *multiple* Issues, but it is possible, for example, if multiple *Bug Reports* are created for the same piece of code.

:::{admonition} Pull Requests don't need self-assignment!
Who is working on a Pull Request is usually obvious.
:::

One approving review should be conducted before any Pull Request is merged.  Please remember to be thorough in your reviews!  Also, remember to spread reviews out so that one person is not doing all of the reviews.

## Notifications

> GitHub notifications can be *extremely useful* for keeping up with what is happening in Xdev!  They are also extremely useful for communicating with other members of the team.  Team members should *direct mention* (@mention) anyone whenever they need that person's input or advice.  When *you* are direct mentioned, you should prioritize responding to those notifications so as not to block progress in the team.  Direct mentions should be responded to with the same priority as requests for reviews.

Reviewer: ```suggestion
GitHub notifications can be *extremely useful* for keeping up with what is happening in Xdev!  They are also extremely useful for communicating with other members of the team.  Team members should *direct mention* (@mention) anyone whenever they need that person's input or advice.  When *you* are directly mentioned, you should prioritize responding to those notifications so as not to block progress in the team.  Direct mentions should be responded to with the same priority as requests for reviews.
```
